### PR TITLE
Support different distribution types for random variables in a new GMF format

### PR DESCRIPTION
**Pull request by eedrummer**
This change supports a new `distribution` property in `Delay`, `Observation`, `Procedure`, `SetAttribute`, `Symptom` and `VitalSign` states. 

For example, an `Observation` of LDL Cholesterol with a Gaussian distribution of results would be:

```json
    "Observation" : {
      "type" : "Observation",
      "category" : "procedure",
      "codes" : [
        {
          "system" : "LOINC",
          "code" : "2089-1",
          "display" : "Cholesterol in LDL"
        }
      ],
      "distribution": {
        "kind": "GAUSSIAN",
        "parameters": {
          "mean": 140,
          "standardDeviation": 15
        }
      },
      "unit": "mg/dL",
    }
```

All states still support the "legacy" format for defining ranges or whatever they call themselves. Current distribution types in this PR are:
* Exact
* Uniform
* Gaussian

The Distribution class should make it easy to add any other distribution types desired.

### CODE REVIEW

**Review by dehall — COMMENTED** [review]
Haven't tested yet but just looking at the code everything seems good, will test it out shortly. One potential idea - currently this is checking for "isLegagyGMF" every time the given state processes, and falls back to legacy behavior if it is. It's only a simple null check so not computationally intensive, but would it simplify things at all to instead "upgrade" the module at initialize time into the latest format?

**Review by hadleynet — CHANGES_REQUESTED** [review]

File: src/main/java/org/mitre/synthea/engine/State.java [hunk 1]
```diff
@@ -1403,9 +1487,18 @@ public void processOnce(Person person, long time) {
...
+        double durationVal;
+        String procedureDurationUnit;
+        if (duration != null) {
+          durationVal = person.rand(duration.low, duration.high);
+          procedureDurationUnit = duration.unit;
+        } else {
+          durationVal = distribution.generate(person);
+          procedureDurationUnit = this.unit;
+        }
+        this.stop = procedure.start + Utilities.convertTime(procedureDurationUnit,
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply src/main/java/org/mitre/synthea/engine/State.java:1500]: Will fix this separately.

**Comment by eedrummer** [comment]
The changes that @hadleynet made look good to me. I like @dehall 's idea about up converting GMF 1.0 states on loading the JSON, as that would allow us to get rid of GMF 1.0 code everywhere but the initialization code, but I don't know if we want to take that on right now.

@jawalonoski I know you don't have a lot of time to look at this, but it would be great to get your thoughts.

File: src/main/java/org/mitre/synthea/engine/State.java [hunk 2]
```diff
@@ -1756,27 +1879,49 @@ protected void initialize(Module module, String name, JsonObject definition) {
...
+      if (distribution != null && !distribution.validate()) {
+        throw new IllegalStateException(
+            String.format("State %s contains an invalid distribution", this.name));
+      }
     }
 
     @Override
     public Symptom clone() {
       Symptom clone = (Symptom) super.clone();
+      clone.symptom = symptom;
```
Comment (hadleynet) [inline src/main/java/org/mitre/synthea/engine/State.java:1891]: Same comment re super.clone doing the work of copying members.

File: src/main/java/org/mitre/synthea/engine/State.java [hunk 3]
```diff
@@ -1313,19 +1378,38 @@ public boolean process(Person person, long time) {
...
+      if (distribution != null && !distribution.validate()) {
+        throw new IllegalStateException(
+            String.format("State %s contains an invalid distribution", this.name));
+      }
+    }
 
     @Override
     public Procedure clone() {
       Procedure clone = (Procedure) super.clone();
+      clone.codes = codes;
```
Comment (hadleynet) [inline src/main/java/org/mitre/synthea/engine/State.java:1401]: I think the super.clone will take care of copying all of the member variables so you can remove lines 1401-1406 inclusive.

**Review by hadleynet — CHANGES_REQUESTED** [review]
Two small changes in your clone methods.

**Comment by eedrummer** [comment]
@hadleynet I cleaned up the clone methods. I think some of that was pulled back in by the rebase I did off of master and not correctly resolving the conflict.

**Review by hadleynet — APPROVED** [review]

**Pull request merged**